File: LeetCode/minimum-path-sum.cpp
// https://leetcode.com/problems/minimum-path-sum/submissions/
class Solution
{
public:
    int minPathSum(vector<vector<int>> &grid)
    {
        vector<vector<int>> path_sum(grid.size(), vector<int>(grid[0].size(), 0));
        path_sum[0][0] = grid[0][0];
        for (int i = 1; i < grid.size(); ++i)
        {
            path_sum[i][0] = path_sum[i - 1][0] + grid[i][0];
        }
        for (int j = 1; j < grid[0].size(); ++j)
        {
            path_sum[0][j] = path_sum[0][j - 1] + grid[0][j];
        }
        for (int i = 1; i < grid.size(); ++i)
        {
            for (int j = 1; j < grid[0].size(); ++j)
            {
                path_sum[i][j] = grid[i][j] + min(path_sum[i - 1][j], path_sum[i][j - 1]);
            }
        }
        return path_sum[grid.size() - 1][grid[0].size() - 1];
    }
};

File: LeetCode/minimum-path-sum.java
// https://leetcode.com/problems/minimum-path-sum/

// Time Complexity: O(mn), Space Complexity: O(mn)
class Solution {
  public int minPathSum(int[][] grid) {
    int rows = grid.length, cols = grid[0].length;
    int[][] dp = new int[rows][cols];

    for (int i = 0; i < rows; ++i) {
      for (int j = 0; j < cols; ++j) {
        // Elements in first row (except the one element in first column).
        if (i == 0 && j != 0) {
          dp[i][j] = grid[i][j] + dp[i][j - 1];
        }
        // Elements in first column (except the one in the first row).
        else if (j == 0 && i != 0) {
          dp[i][j] = grid[i][j] + dp[i - 1][j];
        }
        // All elements except ones in first row or first column.
        else if (i != 0 && j != 0) {
          dp[i][j] = grid[i][j] + Math.min(dp[i - 1][j], dp[i][j - 1]);
        }
        // Element at (0, 0).
        else {
          dp[i][j] = grid[i][j];
        }
      }
    }

    return dp[rows - 1][cols - 1];
  }
}

// Time Complexity: O(mn), Space Complexity: O(1)

class Solution {
  public int minPathSum(int[][] grid) {
    int rows = grid.length, cols = grid[0].length;

    for (int i = 0; i < rows; ++i) {
      for (int j = 0; j < cols; ++j) {
        // Elements in first row (except the one element in first column).
        if (i == 0 && j != 0) {
          grid[i][j] = grid[i][j] + grid[i][j - 1];
        }
        // Elements in first column (except the one in the first row).
        else if (j == 0 && i != 0) {
          grid[i][j] = grid[i][j] + grid[i - 1][j];
        }
        // All elements except ones in first row or first column.
        else if (i != 0 && j != 0) {
          grid[i][j] = grid[i][j] + Math.min(grid[i - 1][j], grid[i][j - 1]);
        }
        // Element at (0, 0).
        // else {
        // grid[i][j] = grid[i][j];
        // }
      }
    }

    return grid[rows - 1][cols - 1];
  }
}